tests/core_test/mels_test.py

```diff
@@ -0,0 +1,196 @@
+    random_key, subkey = jax.random.split(random_key)
+    keys = jax.random.split(subkey, num=batch_size)
+    fake_batch = jnp.zeros(shape=(batch_size, env.observation_size))
+    init_variables = jax.vmap(policy_network.init)(keys, fake_batch)
+
+    # Create the initial environment states. Note that there are batch_size * num_evals
+    # environments since each controller must be evaluated num_evals times.
+    keys = jax.random.split(random_key, batch_size * num_evals + 1)
+    random_key = keys[0]
+    subkeys = keys[1:]
+    reset_fn = jax.jit(jax.vmap(env.reset))
+    init_states = reset_fn(subkeys)
+
+    # Define the function to play a step with the policy in the environment
+    def play_step_fn(
+        env_state: EnvState,
+        policy_params: Params,
+        random_key: RNGKey,
+    ) -> Tuple[EnvState, Params, RNGKey, QDTransition]:
+        """Play an environment step and return the updated state and the
+        transition."""
+
+        actions = policy_network.apply(policy_params, env_state.obs)
+
+        state_desc = env_state.info["state_descriptor"]
+        next_state = env.step(env_state, actions)
+
+        transition = QDTransition(
+            obs=env_state.obs,
+            next_obs=next_state.obs,
+            rewards=next_state.reward,
+            dones=next_state.done,
+            actions=actions,
+            truncations=next_state.info["truncation"],
+            state_desc=state_desc,
+            next_state_desc=next_state.info["state_descriptor"],
+        )
+
+        return next_state, policy_params, random_key, transition
+
+    # Prepare the scoring function
+    bd_extraction_fn = environments.behavior_descriptor_extractor[env_name]
+
+    def scoring_fn(
+        policies_params: Genotype,
+        random_key: RNGKey,
+    ) -> Tuple[Fitness, Descriptor, ExtraScores, RNGKey]:
+        # Repeat each policy so that it can be evaluated num_evals times.
+        policies_params = jax.tree_util.tree_map(
+            lambda x: jnp.repeat(x, num_evals, axis=0), policies_params
+        )
+
+        # Call the Brax scoring function, taking into account previously defined
+        # inputs like `init_states`.
+        fitnesses, descriptors, extra_scores, random_key = scoring_function_brax_envs(
+            policies_params=policies_params,
+            random_key=random_key,
+            init_states=init_states,
+            episode_length=episode_length,
+            play_step_fn=play_step_fn,
+            behavior_descriptor_extractor=bd_extraction_fn,
+        )
+
+        # Reshape results for ME-LS.
+        extra_scores["transitions"] = jax.tree_util.tree_map(
+            lambda x: x.reshape((batch_size, num_evals, *x.shape[1:])),
+            extra_scores["transitions"],
+        )
+        return (
+            fitnesses.reshape((batch_size, num_evals)),
+            descriptors.reshape((batch_size, num_evals, -1)),
+            extra_scores,
+            random_key,
+        )
```

Comment:
I discussed the way the scoring function is defined here with @limbryan,  and we think that `scoring_fn` should be as general as possible. For example, we should define a `scoring_fn` that could be passed to `MELS` but also to the standard `MAPElites` class.

To do that, I suggest that you:
- pass the standard `scoring_fn` to the `MELS` class.
- and in the `MELS` class, you override `update(...)` with a code that does a `vmap` on the scoring fn to evaluate it on several individuals. Here is an approximate suggestion on how to change `scoring_function` call in the `update(...)` method.
```python
...
random_key, *subkeys = jax.random.split(random_key, self.num_evals + 1)  # num_evals can be an attribute of the `MELS` class
fitnesses, descriptors, extra_scores, _ = jax.vmap(self.scoring_function, in_axes=(None, 0))(genotypes, jnp.asarray(subkeys))
...  #  adding to container and so on...
```

Reply:
I think it makes sense to have a scoring_fn that is the same as in MAPElites, but one issue I see is that the brax scoring function takes in a set of `init_states`. Using the vmap approach, it seems the `init_states` would be repeated across all the evals, thus resulting in 0 spread (IIRC the initial state is the only non-deterministic component in these environments). 

One potential idea (albeit a backwards-incompatible one) is that perhaps `scoring_function_brax_envs` should actually be taking in a series of random keys, one for each controller it is evaluating. That way, if you want to create different environments, you can pass in a ton of random keys, and if you don't want stochasticity, you pass in the same random key repeated several times.

Reply:
Not off the top of my head (the tutorials need some updating in general but haven't had the time to get around it) but ```env.reset``` or ```partial(env.reset)``` should work as the ```play_reset_fn```

Reply:
@Lookatator @limbryan I finished implementing this change! I added a `multi_eval_scoring_function` in `MELS.__init__`. The scoring function now uses `reset_based_scoring_function_brax_envs`

Reply:
Great! Thank you @btjanaka!